### PR TITLE
Who said it?

### PR DESCRIPTION
**Pull request by vsporeddy**
Who said it? Players guess the correct OP for a random quote from the server

co-authored by @mcab

### CODE REVIEW

**Comment by mcab** [comment]
@ddcha3 How can we make the polling work with `asyncio` without having to use a spinlock?

File: games/whosaidit/whosaidit.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        return guess.author in players and \
+                guess.channel == bot.get_channel(config["game_channel"]) and \
+                guess.content.isnumeric() and \
+                int(guess.content) in range(1, len(choices) + 1)
+
+    guesses = OrderedDict()
+    while len(guesses) < len(players):
+        guess = await bot.wait_for("message", check=check_guess)
+        if guess.author.id not in guesses:
+            guesses[guess.author.id] = int(guess.content)
```
Comment (mcab) [inline games/whosaidit/whosaidit.py:94]: This specifically provides an issue if too many messages come in at once.

We tried `asyncio.wait_for()`ing this function, with a timeout if the condition isn't met. However, the async/await primitives being used don't match up, hence no guesses actually get recorded.

File: games/whosaidit/whosaidit.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    answer = list(choices.keys())[list(choices.values()).index(message_to_guess.author)]
+
+    await display(ctx, message_to_guess, choices)
+    guesses = await get_choice(bot, ctx, choices, players)
+    winners = []
+    for player, guess in guesses.items():
+        if guess == answer:
+            winners.append(bot.get_user(player))
+            await ctx.send(f"This quote was from {message_to_guess.author.name} on <t:{int(message_to_guess.created_at.timestamp())}:d>!")
+            return winners
```
Comment (mcab) [inline games/whosaidit/whosaidit.py:80]: This results in a bug.

If we do not have _any_ correct guesses, `winners` is never returned, and we end in a bad state. A potential way is to return empty winners, but we can probably refactor for better logic here.

```suggestion
    for player, guess in guesses.items():
        if guess == answer:
            winners.append(bot.get_user(player))
            await ctx.send(f"This quote was from {message_to_guess.author.name} on <t:{int(message_to_guess.created_at.timestamp())}:d>!")
            return winners
    return winners
```

Comment (vsporeddy) [reply games/whosaidit/whosaidit.py]: Reworked to return empty winners, that's how other non zero-sum chowderbot games work for coin/cgr calculations

**Pull request merged**